postgis/lwgeom_remove_irrelevant_points_for_view.c

```diff
@@ -0,0 +1,327 @@
+	vxall = 0;
+	vyall = 0;
+    for (i = 0; i < npoints; i++) {
+
+		// get current point [i]
+		getPoint4d_p(points, i, &p);
+		vx = encodeToBits(p.x, xmin, xmax);
+		vy = encodeToBits(p.y, ymin, ymax);
+
+		// get subsequent point [i+1]
+		next = i + 1;
+		if (next == npoints) {
+			if (closed) next = points -> npoints-1; // for polygons, same as original start point which may be overwritten
+			else next = i;  // for linestrings reuse end point
+		}
+		getPoint4d_p(points, next, &p1);
+		vx1 = encodeToBits(p1.x, xmin, xmax);
+		vy1 = encodeToBits(p1.y, ymin, ymax);
+
+		sameX = vx == vx1 && vx == vx0;
+		sameY = vy == vy1 && vy == vy0;
+	    insideX = vx == 0x02;
+	    insideY = vy == 0x02;
+	    inside = insideX && insideY;
+
+	    skip = sameX && sameY && !inside;	// three consecutive points in same outside quarter, leave out central one
+	    skip |= sameX && !insideX;			// three consecutive points in same outside area (left or right), leave out central one
+	    skip |= sameY && !insideY;			// three consecutive points in same outside area (top or buttom), leave out central one
+
+        if (skip) continue;
+
+        // save current point at [w <= i]
+        ptarray_set_point4d(points, w++, &p);
+	    vx0 = vx;
+	    vy0 = vy;
+        vxall |= vx;
+        vyall |= vy;
+    }
+
+    if (closed && w > 0) {
+	    // re-add first new point at the end if closed
+		getPoint4d_p(points, 0, &p);
+        ptarray_set_point4d(points, w++, &p);
+    }
+
+    // eval empty cases
+    clear = w < minpoints; 		// too less points
+    clear |= vxall == 0x01;		// completely left outside
+    clear |= vxall == 0x04;		// completely right outside
+    clear |= vyall == 0x01;		// completely top outside
+    clear |= vyall == 0x04;		// completely bottom outside
+    if (clear) w = 0;
+
+    points->npoints = w;
+}
+
+// ===============================================================================
+// remove points that are irrelevant for rendering the geometry within
+// a view specified by rectangular bounds.
+// 2D-(MULTI)POLYGONs and (MULTI)LINESTRINGs are evaluated, others keep untouched.
+// ===============================================================================
+PG_FUNCTION_INFO_V1(ST_RemoveIrrelevantPointsForView);
+Datum ST_RemoveIrrelevantPointsForView(PG_FUNCTION_ARGS) {
+
+    unsigned int i, j, iw, jw;
+
+    // gserialized logic see for example in /postgis/lwgeom_functions_basic.c,
+    // type definitions see /liblwgeom/liblwgeom.h(.in)
+
+    GSERIALIZED *serialized_in;
+    GSERIALIZED *serialized_out;
+
+    LWGEOM *geom;
+    GBOX *bbox;
+
+    // geom input check
+    if (PG_GETARG_POINTER(0) == NULL) {
+        PG_RETURN_NULL();
+    }
+
+    serialized_in = (GSERIALIZED *)PG_DETOAST_DATUM_COPY(PG_GETARG_DATUM(0));
+
+    // box input check
+    if (PG_GETARG_POINTER(1) == NULL) {
+		// no BBOX given, leave untouched
+		PG_RETURN_POINTER(serialized_in);
+    }
+
+    // dimension check (only 2d geometry types are supported yet)
```

Comment:
I think @lnicola  complained about this already.  I think this should be > 2 or != 2.  We don't have any 1 ndim geometries to my knowledge so this is not protecting against anything.

Reply:
Since it is irrelevant I removed this check (and added two tests).